deployment/ansible/scripts/topology.py

```diff
@@ -40,51 +45,132 @@ def hosts_to_peers(lines):
 def read_inventories(path):
     with open(path, "r") as f:
         lines = f.readlines()
-    key = lines[0].strip()
-    inventories = {key: []}
-    for _line in lines[1:]:
+    inventories = defaultdict(list)
+    for _line in lines:
         line = _line.strip()
         if not line.startswith("["):
             inventories[key].append(line)
         else:
             key = line
-            inventories[key] = []
-    print(inventories)
-    peers = hosts_to_peers(inventories["[nodes]"])
-    return peers
+    return inventories
 
 
-def barbell_topology(_inventories):
-    inventories = deepcopy(_inventories)
-    random.shuffle(inventories)
-    n = len(inventories)
-    topology = {peer: [inventories[(i + 1) % n]]
-                for i, peer in enumerate(inventories)}
-    return topology
+def barbell_topology(_peers):
+    peers = deepcopy(_peers)
+    random.seed(5566)
+    random.shuffle(peers)
+    n = len(peers)
+    topology = [{
+        'peer': peer,
+        'connect_to': [peers[(i + 1) % n]],
+        'shard_id': int(i/3),
+        'is_broadcasting': i == 0
+    } for i, peer in enumerate(peers)]
 
+    return topology
 
-def topology_to_yaml(topology):
-    results = {}
-    for me, others in topology.items():
-        commands = [
-            f"docker exec -t {me.container_name} sh -c './sharding-p2p-poc -loglevel=DEBUG -client addpeer {peer.public_ip} {peer.listen_port} {peer.seed}'"
-            for peer in others
-        ]
-        if me.host in results:
-            results[me.host].extend(commands)
-        else:
-            results[me.host] = commands
 
-    return yaml.dump(results, default_flow_style=False, width=1000)
+def groupby_host(commands, hosts):
+    groups = {host: [] for host in hosts}
```

Comment:
seems `defaultdict` can be used here

Reply:
Ah, I intentionally leave the empty lists here. Otherwise, in the ansible script, those servers would fail to find a key if using defaultdicts.